security/advancedtls/crl.go

```diff
@@ -0,0 +1,498 @@
+	RevocationRevoked
+)
+
+func (s RevocationStatus) String() string {
+	return [...]string{"RevocationUndetermined", "RevocationUnrevoked", "RevocationRevoked"}[s]
+}
+
+// certificateListExt contains a pkix.CertificateList and parsed
+// extensions that aren't provided by the golang CRL parser.
+type certificateListExt struct {
+	CertList *pkix.CertificateList
+	// RFC5280, 5.2.1, all conforming CRLs must have a AKID with the ID method
+	AuthorityKeyID []byte
+}
+
+const tagDirectoryName = 4
+
+var (
+	// RFC5280, 5.2.4 id-ce-deltaCRLIndicator OBJECT IDENTIFIER ::= { id-ce 27 }
+	oidDeltaCRLIndicator = asn1.ObjectIdentifier{2, 5, 29, 27}
+	// RFC5280, 5.2.5 id-ce-issuingDistributionPoint OBJECT IDENTIFIER ::= { id-ce 28 }
+	oidIssuingDistributionPoint = asn1.ObjectIdentifier{2, 5, 29, 28}
+	// RFC5280, 5.3.3 id-ce-certificateIssuer   OBJECT IDENTIFIER ::= { id-ce 29 }
+	oidCertificateIssuer = asn1.ObjectIdentifier{2, 5, 29, 29}
+	// RFC5290, 4.2.1.1 id-ce-authorityKeyIdentifier OBJECT IDENTIFIER ::=  { id-ce 35 }
+	oidAuthorityKeyIdentifier = asn1.ObjectIdentifier{2, 5, 29, 35}
+)
+
+// x509NameHash implements the OpenSSL X509_NAME_hash function for hashed directory lookups.
+func x509NameHash(r pkix.RDNSequence) string {
+	var canonBytes []byte
+	// First canonicalize all the strings
+	for _, rdnSet := range r {
+		for i, rdn := range rdnSet {
+			value, ok := rdn.Value.(string)
+			if !ok {
+				continue
+			}
+			// OpenSSL trims all whitespace, does a tolower, and removes extra spaces between words
+			// Implemented in x509_name_canon in OpenSSL
+			canonStr := strings.Join(strings.Fields(
+				strings.TrimSpace(strings.ToLower(value))), " ")
+			// Then it changes everything to UTF8 strings
+			rdnSet[i].Value = asn1.RawValue{Tag: asn1.TagUTF8String, Bytes: []byte(canonStr)}
+
+		}
+	}
+
+	// Finally, OpenSSL drops the initial sequence tag
+	// So, we marshal all the RDNs separately instead of as a group
+	for _, canonRdn := range r {
+		b, err := asn1.Marshal(canonRdn)
+		if err != nil {
+			grpclog.Print(err)
+		}
+		canonBytes = append(canonBytes, b...)
+	}
+
+	issuerHash := sha1.Sum(canonBytes)
+	// Openssl takes the first 4 bytes and encodes them as a little endian
+	// uint32 and then uses the hex to make the file name
+	// In C++, this would be:
+	// (((unsigned long)md[0]) | ((unsigned long)md[1] << 8L) |
+	// ((unsigned long)md[2] << 16L) | ((unsigned long)md[3] << 24L)
+	// ) & 0xffffffffL;
+	fileHash := binary.LittleEndian.Uint32(issuerHash[0:4])
+	return fmt.Sprintf("%08x", fileHash)
+}
+
+// CheckRevocation checks the connection for revoked certificates based on RFC5280.
+// This implementation has the following major limitations:
+// 	Indirect CRL files are not supported
+// 	File based lookup only using the X509_LOOKUP_hash_dir format
+// 	OnlySomeReasons is unsupported
+// 	Delta CRL files are unsupported
+// 	Certificate cRLDistributionPoint must be URLs, but are then ignored and converted into a file path
```

Comment:
cRL -> CRL

Reply:
done, cRL is the capitalization from the RFC but x509.Certificate has leading caps.